Rebase to latest master branch

BlGene opened the pull request:
Hi @weiliu89 ,
This is a re-base of all fcn layers to the latest caffe master, plus the solver modification. Each added layer is squashed into one commit. The modified pooling layer was re-named to as binpooling layer, however it should be able to replace this binpooling layer with a SPP pooling layer. It would be great if you could have a look at this.
BR, Max

weiliu89 commented:
@BlGene Thank you for rebasing.

For the normalization layer, can you replace it with this commit?

https://github.com/BVLC/caffe/commit/89380f113fdbf39291a309d6bd7b9d3cd70a7940

Also, I am not sure if we need BinPooling layer because it can be replaced by SPP layer or Pooling layer.

BlGene commented:
Hi @weiliu89,

1. Updated normalization layer accordint to BVLC@89380f1, and added test_normalize_layer.cpp which I forgot previously.
2. Removed BinPooling layer, I did mention that this layer was most probably redundant ;).
3. This sill has the basic crop layer, hopefully #3570 will be merged soon, at which point the commit can be replaced.

weiliu89 commented:
Hi @BlGene,

What is the purpose of this pull request? Are you planning to merge with BVLC master branch?

BlGene commented:
Hi Wei,

I started the re-base because I wanted compatibility with the new cuDNN version. I posted the re-base to avoid duplicate work, in case you or anybody else were planning to do the same.

Merging into BVLC master would be nice, but it depends on how interested they are in merging the required layers. The crop layer will be merged, but there might be little interest in merging the Normalization Layer and the UnPooling Layer, although these seem generically useful. The other layers and the solver modification are not essential. 
@longjon  care to comment?

Even if this PR doesn't get merged into master I hope it can be a up-to-date basis for segmentation work.

BR, Max

weiliu89 commented:
Hi Max,

Awesome. Thanks for explanation!

Since I didn't follow Caffe for a period of time, what happened to CropLayer? I notice that it uses the offset_height and offset_width parameters now, instead of deriving automatically from the network structure.

BTW. It would be great if there is an exemplar .prototxt to put all these together?

Best,
Wei

BlGene commented:
Parameters can still be derived automatically using PR #3613, but as a separate pre-processing step, instead of during run-time of the net. IIRC offsets were 0 for your net.
For the example net only the BinPooling->SPP and CropLayer changes would need to be made, i'll try posting one later today.
BR, Max

BlGene commented:
@weiliu89 
Updated to using master's SPP and crop layer. In case it works as advertised for you, maybe you could consider merging.

For parseNet to function on master we would still be missing:
1. Normalization (No suitable PR yet.)
2. UnPooling . (PR #3601 should work for this, haven't checked though).
